mitmproxy/addons/maplocal.py

```diff
@@ -1,52 +1,86 @@
 import mimetypes
 import re
 import typing
-import urllib
 from pathlib import Path
+
 from werkzeug.security import safe_join
 
-from mitmproxy import ctx, exceptions, http
-from mitmproxy.addons.modifyheaders import parse_modify_spec, ModifySpec
+from mitmproxy import ctx, exceptions, flowfilter, http
+from mitmproxy.addons.modifyheaders import parse_spec
+
+
+class MapLocalSpec(typing.NamedTuple):
+    matches: flowfilter.TFilter
+    regex: str
+    local_path: Path
+
+
+def parse_map_local_spec(option: str) -> MapLocalSpec:
+    filter, regex, replacement = parse_spec(option)
+
+    try:
+        re.compile(regex)
+    except re.error as e:
+        raise ValueError(f"Invalid regular expression {regex!r} ({e})")
+
+    try:
+        path = Path(replacement).expanduser().resolve(strict=True)
+    except FileNotFoundError as e:
+        raise ValueError(f"Invalid file path: {replacement} ({e})")
+
+    return MapLocalSpec(filter, regex, path)
 
 
 def get_mime_type(file_path: str) -> str:
     mimetype = (
-        mimetypes.guess_type(file_path)[0]
-        or "application/octet-stream"
+            mimetypes.guess_type(file_path)[0]
+            or "application/octet-stream"
     )
     return mimetype
 
 
-def file_candidates(url: str, base_path: str) -> typing.List[Path]:
-    candidates = []
-    parsed_url = urllib.parse.urlparse(url)
-    path_components = parsed_url.path.lstrip("/").split("/")
-    filename = path_components.pop()
+def _safe_path_join(root: Path, untrusted: str) -> Path:
+    """Join a Path element with an untrusted str.
 
-    # todo: we may want to consider other filenames such as index.htm)
-    if not filename:
-        filename = 'index.html'
+    This is just a convenience wrapper for werkzeug's safe_join."""
+    untrusted_parts = Path(untrusted).parts
+    joined = safe_join(
+        root.as_posix(),
+        *untrusted_parts
+    )
+    if joined is None:
+        raise ValueError("Untrusted paths.")
+    return Path(joined)
 
-    # construct all possible paths
-    while True:
-        components_with_filename = tuple(path_components + [filename])
-        candidate_path = safe_join(base_path, *components_with_filename)
-        if candidate_path:
-            candidates.append(
-                Path(candidate_path)
-            )
 
-        if not path_components:
-            break
+def file_candidates(url: str, spec: MapLocalSpec) -> typing.List[Path]:
+    candidates = []
 
-        path_components.pop()
+    m = re.search(spec.regex, url)
+    assert m
```

Comment:
`assert m` instead of a `if m:` block?

Reply:
We can assert here because we already check for a match in `request`, but I'll quickly refactor this.